### PR TITLE
Update NNCF Dataloader in Anomaly Task

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
### Summary

- Replicate changes in https://github.com/openvinotoolkit/anomalib/pull/2286 as it leads to higher performance.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/anomaly/openvino_model.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ def check_if_quantized(model: openvino.Model) -> bool:
             msg = "Model is already optimized by PTQ"
             raise RuntimeError(msg)
 
-        val_dataset = data_module.train_dataloader()
+        val_dataset = data_module.val_dataloader()
```
Comment (sovrasov) [inline src/otx/algo/anomaly/openvino_model.py:176]: Does that change lead to increased accuracy on val dataset?
We use train subset for quantization of other OTX models to prevent data leak.

Comment (ashwinvaidya17) [reply src/otx/algo/anomaly/openvino_model.py:176]: At least in anomalib I am using the test dataset for checking the performance.
In OTX, these are the commands I used for testing. Below you can see the testing performance due to the change between val/test dataset. I am assuming that having some anomalous images when optimizing has a positive effect on the performance.

# `hazelnut_large`

```bash
otx train --config src/otx/recipe/anomaly_classification/stfpm.yaml --data_root ./datasets/otx/v2/anomaly/mvtec/hazelnut_large --work_dir otx-stfpm 
```

```bash
otx test --checkpoint /home/ashwin/ote/otx-stfpm/.latest/train/best_checkpoint.ckpt --config src/otx/recipe/anomaly_classification/stfpm.yaml --data_root ./datasets/otx/v2/anomaly/mvtec/hazelnut_large --work_dir otx-stfpm
```

```bash
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │    0.9957143068313599     │
│       image_F1Score       │    0.9718309640884399     │
│      test/data_time       │    0.1736757755279541     │
│      test/iter_time       │    0.1937289535999298     │
└───────────────────────────┴───────────────────────────┘
```

```bash
otx export --work_dir otx-stfpm --checkpoint otx-stfpm/.latest/train/best_checkpoint.ckpt
```

```bash
otx optimize --work_dir otx-stfpm --checkpoint /home/ashwin/ote/otx-stfpm/20240903_073939/exported_model.xml --data_root ./datasets/otx/v2/anomaly/mvtec/hazelnut_large
```

## val_dataloader
```bash
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │    0.9980356693267822     │
│       image_F1Score       │    0.9857142567634583     │
│      test/data_time       │   0.0028419645968824625   │
│      test/iter_time       │   0.052088286727666855    │
└───────────────────────────┴───────────────────────────┘
```
## train_dataloader

```bash
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        image_AUROC        │            1.0            │
│       image_F1Score       │            0.0            │
│      test/data_time       │    0.00306075531989336    │
│      test/iter_time       │   0.054295849055051804    │
└───────────────────────────┴───────────────────────────┘
```

Comment (sovrasov) [reply src/otx/algo/anomaly/openvino_model.py:176]: Thanks for the explanation. That fix is ok as a workaround.

**Review by sovrasov — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]

**Comment by harimkang** [comment]
I have a question, don't some of the models in the anomaly task have no validation, and in certain cases do use them without validation data?

**Comment by ashwinvaidya17** [comment]
> I have a question, don't some of the models in the anomaly task have no validation, and in certain cases do use them without validation data?

We use validation set in Anomalib for finding the normalization thresholds. It is also used by few-shot models for reference images. If validation set is not used in anomaly tasks in OTX then that is a bug in the workflow.
